libraries/rush-lib/src/logic/operations/BuildPlanPlugin.ts

```diff
@@ -0,0 +1,298 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import type { ITerminal } from '@rushstack/terminal';
+import type {
+  IExecuteOperationsContext,
+  IPhasedCommandPlugin,
+  PhasedCommandHooks
+} from '../../pluginFramework/PhasedCommandHooks';
+import type { Operation } from './Operation';
+import { clusterOperations, type IOperationBuildCacheContext } from './CacheableOperationPlugin';
+import { DisjointSet } from '../cobuild/DisjointSet';
+import type { IOperationExecutionResult } from './IOperationExecutionResult';
+import { RushConstants } from '../RushConstants';
+import type { RushProjectConfiguration } from '../../api/RushProjectConfiguration';
+
+const PLUGIN_NAME: 'BuildPlanPlugin' = 'BuildPlanPlugin';
+
+interface IBuildPlanOperationCacheContext {
+  cacheDisabledReason: IOperationBuildCacheContext['cacheDisabledReason'];
+}
+
+interface ICobuildPlan {
+  summary: {
+    maxWidth: number;
+    maxDepth: number;
+    numberOfNodesPerDepth: number[];
+  };
+  executionPlan: Operation[];
+  clusters: Set<Operation>[];
+  buildCacheByOperation: Map<Operation, IBuildPlanOperationCacheContext>;
+  clusterByOperation: Map<Operation, Set<Operation>>;
+}
+
+export class BuildPlanPlugin implements IPhasedCommandPlugin {
+  private readonly _terminal: ITerminal;
+
+  public constructor(terminal: ITerminal) {
+    this._terminal = terminal;
+  }
+
+  public apply(hooks: PhasedCommandHooks): void {
+    const terminal: ITerminal = this._terminal;
+    hooks.beforeExecuteOperations.tapPromise(PLUGIN_NAME, createBuildPlan);
+
+    async function createBuildPlan(
+      recordByOperation: Map<Operation, IOperationExecutionResult>,
+      context: IExecuteOperationsContext
+    ): Promise<void> {
+      const { projectConfigurations, projectChangeAnalyzer } = context;
+      const disjointSet: DisjointSet<Operation> = new DisjointSet<Operation>();
+      const operations: IterableIterator<Operation> = recordByOperation.keys();
+      for (const operation of operations) {
+        disjointSet.add(operation);
+      }
+      const buildCacheByOperation: Map<Operation, IBuildPlanOperationCacheContext> = new Map<
+        Operation,
+        IBuildPlanOperationCacheContext
+      >();
+      for (const operation of operations) {
+        const { associatedProject, associatedPhase } = operation;
+        if (associatedProject && associatedPhase) {
+          const projectConfiguration: RushProjectConfiguration | undefined =
+            projectConfigurations.get(associatedProject);
+          const fileHashes: Map<string, string> | undefined =
+            await projectChangeAnalyzer._tryGetProjectDependenciesAsync(associatedProject, terminal);
+          const cacheDisabledReason: string | undefined = projectConfiguration
+            ? projectConfiguration.getCacheDisabledReason(fileHashes!.keys(), associatedPhase.name, operation)
```

Comment:
```suggestion
            ? projectConfiguration.getCacheDisabledReason(fileHashes!.keys(), associatedPhase.name, operation?.runner?.isNoOp)
```
That avoids a lot of coupling.